volume/exportfs.go

```diff
@@ -0,0 +1,173 @@
+// limitations under the License.
+
+package volume
+
+import (
+	"archive/tar"
+	"io"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/zenoss/glog"
+)
+
+// ExportDirectory recursively writes its contents into a tar Writer.
+func ExportDirectory(tarfile *tar.Writer, path, name string) error {
+	dir, err := os.Open(path)
+	if err != nil {
+		glog.Errorf("Could not open %s: %s", path, err)
+		return err
+	}
+	defer dir.Close()
+	fstat, err := dir.Stat()
+	if err != nil {
+		glog.Errorf("Could not stat %s: %s", path, err)
+		return err
+	}
+	header, err := getHeader(name, "", fstat)
+	if err != nil {
+		return err
+	}
+	if err := tarfile.WriteHeader(header); err != nil {
+		glog.Errorf("Could not write header for directory %s: %s", path, err)
+		return err
+	}
+	files, err := dir.Readdir(0)
+	if err != nil {
+		glog.Errorf("Could not list directory for %s: %s", path, err)
+		return err
+	}
+	for _, finfo := range files {
+		fullpath, relpath := filepath.Join(path, finfo.Name()), filepath.Join(name, finfo.Name())
+		if finfo.IsDir() {
+			if err := ExportDirectory(tarfile, fullpath, relpath); err != nil {
+				return err
+			}
+		} else {
+			if err := ExportFile(tarfile, fullpath, relpath); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ExportFile writes a file into a tar Writer.
+func ExportFile(tarfile *tar.Writer, path, name string) error {
+	file, err := os.Open(path)
+	if err != nil {
+		glog.Errorf("Could not open %s: %s", path, err)
+		return err
+	}
+	defer file.Close()
+	fstat, err := file.Stat()
+	if err != nil {
+		glog.Errorf("Could not stat %s: %s", path, err)
+		return err
+	}
+
+	link, err := filepath.EvalSymlinks(path)
+	if err != nil {
+		glog.Errorf("Could not check link for %s: %s", path, err)
+		return err
+	}
+	header, err := getHeader(name, link, fstat)
+	if err != nil {
+		glog.Errorf("Could not create file header %s: %s", path, err)
+		return err
+	}
+	if err := tarfile.WriteHeader(header); err != nil {
+		glog.Errorf("Could not write file header %s: %s", path, err)
+		return err
+	}
+	if link == path {
+		if _, err := io.Copy(tarfile, file); err != nil {
+			glog.Errorf("Could not write file %s: %s", path, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// ImportArchive reads from a tar Reader and writes the contents into a path
+// preserving file permissions and ownership.
+func ImportArchive(tarfile *tar.Reader, path string) error {
+	for {
+		header, err := tarfile.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			glog.Errorf("Could not import archive to %s: %s", path, err)
+			return err
+		}
+		if err := ImportArchiveHeader(header, tarfile, path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ImportArchiveHeader imports a tarfile header to a particular path
+func ImportArchiveHeader(header *tar.Header, reader io.Reader, path string) error {
+	filename := filepath.Join(path, header.Name)
+	switch header.Typeflag {
+	case tar.TypeDir:
```

Comment:
Add a comment here saying that, yeah, we aren't handling all types, but these are the only ones we will write.